### ISSUE TITLE
why not use jackson json to speed up performance?

### ISSUE THREAD

**Issue by lxbzmy** [searchbox-io/Jest#45]
why not use jackson json to speed up performance?

**Comment by bobbyhubbard** [comment]
Is someone working on this? I might be able to take it if not

**Comment by kramer** [comment]
@bobbyhubbard No, nobody is working on this currently, you can pick it up if you wish. Since it's kind of a big change it'd be nice to hear your rough attack plan beforehand or to see your progress on small increments to give early feedback :).

**Comment by bobbyhubbard** [comment]
Sure. I'll take a little deeper look

**Comment by ctrimble** [comment]
I took a look at what would be required to replace `Gson` with `ObjectMapper` last night and it looks like the changes will be significant.  In particular, `JsonObject` can be found almost throughout the code base, including a lot of manual mapping type code.

I would be willing to help get this moving, but the way things stand, it would be very hard to produce a super clean migration path.  In the best case, perhaps a version to step through where the `JsonObject` was generic could be done, with the old commands moved to a new module.  This would probably also necessitate use of Java8, to abstract things like testing for errors in the parsed body in `JestResult`.

@kramer Would the project be willing to take a change like this?  I could workup a PR for the search API, but I don't want to put in a ton of effort and end up forking your library.

**Comment by ctrimble** [comment]
@bobbyhubbard did your investigation into using Jackson with Jest produce any results?

**Comment by kramer** [comment]
@ctrimble At this point I'm *conflicted* about this switch considering pros & cons. I don't consider "speed" as a deciding factor because (i) after all each json operation is paired up with a network operation which is much more "expensive" than doing (de)serialization and (ii) a simple google search forms the opinion that recently gson caught up with jackson in speed for most use cases. Then we are left with some parse case pros for jackson like: better (?) annotation support, partial doc/class support and maybe better handling of cases like the one @daniilyar mentioned; and the biggest con is the huge breaking change for the public api and the actual effort needed to do the switch. If those are all we have and we are discussing complete switch to jackson then that is where I'm conflicted: is it really worth it?. If there was a way to keep gson and have jackson in addition then I'd be all for it.

I am very much open to discussion on this.



p.s.: 
> necessitate use of Java8

I don't think android module would like that.

**Comment by ctrimble** [comment]
@kramer for me, and I would assume others, this has nothing to do with performance.  If performance was that big of a deal, it would be worth purpose building something.

What I am after is clean call sites.  The way things stand, I end up with lots of `JsonObject` based code in my project.  It is getting to the point that I am ready to fork `Jest`, just to get that worked out.  I would much rather start contributing here and help everyone.

I am sure that some kind of step through version could be produced, allowing people to half upgrade.  I know I would need that.  I have way too much code at this point to make that lift in one shot.

As for `Java8`, that would not be a necessity, but some strategies will be needed to make a reasonably smooth transition.  Lambdas just make that kind of code a non issue.

Nirvana would be a client api that looked more like this:

```
SearchResult result = client.execute(action);
if( !result.isSucceeded() ) {
  // fail
}
List<MyType<SubType>> hits = result.hits().as(GenericType<MyType<SubType>>(){});
```

**Comment by kramer** [comment]
> step through version

Could you explain that in more detail please? My train of thought on this: it's not like we have constant critical updates; so people can just continue using the old version until they are ready to make the change.

**Comment by ctrimble** [comment]
I have lots of code on my hands, so I would want an upgrade to go something like this:

1. update maven dependency version.
2. bulk update all package Jest related package names.
3. upgrade/add some code against the updated API.

At the same time, I would probably be willing to do a bulk port in this situation.  It really depends on where the refactoring lands.  More than half of my code dealing with `Jest` is wrangling `JsonObject`.  The code is ugly and error prone.  It would be worth the effort.

Would you be willing to look at a throw away PR for the search API?  I could knock something like that together without too much effort, just to give us something concrete to talk about.

**Comment by kramer** [comment]
> Would you be willing to look at a throw away PR for the search API?

@ctrimble Yes, of course.

**Comment by nicoruti** [comment]
+1 for injecting an own serializer through an interface.

I switched to Jest from the official ES client, where the serialization is left to the user. Using the ES client, I created my (annotated) Objects to fit an ObjectMapper / Jackson. Now moving to Jest (with built-in Gson) would mean to change the Objects, which is not possible.

For compatibility with all projects, you could provide a method like <code>JestResponse#getSource(): JsonElement</code> or even <code>JestResponse#getSource(): String</code>.

I'm doing the following workaround at the moment:
```
JsonObject source = response.getSourceAsObject(JsonObject.class);
source.remove(JestResult.ES_METADATA_ID); // optional
MyClass result = objectMapper.readValue(source.toString(), MyClass.class);
```
Although it's not very efficient, it works.

**Comment by ctrimble** [comment]
I got swamped while working on an example PR for this and I don't think I will have time to get back to it soon.

**Comment by nicoruti** [comment]
I looked into the code; it makes heavy use of Gson's JsonObject / JsonElement / JsonArray class. Replacing Gson would be a major refactoring. Nevertheless, I created a PR (see #249 ) which adds a method to get the result as a plain string, such that developers can parse it themselves with any preferred mapper. In my opinion a missing feature.

**Comment by raunak** [comment]
@ctrimble can you push the work you did? I can attempt to complete the remaining parts. 

Most frameworks out there use jackson for parsing json (in my case play framework), and it really annoys me that you end up importing different libraries into a project designed to do the same thing. For example, I now have POJO's with jackson and gson annotations. It might just be me, but I don't find that ok.

**Comment by ctrimble** [comment]
@raunak it isn't in a compiling state, but here you go.  https://github.com/ctrimble/Jest/tree/experiment_object-mapper

**Comment by antonsarov** [comment]
+1 @raunak

It is not just the performance factor. I have Jackson annotations on POJOs (e.g. @JsonManagedReference) which help me a lot for the internal REST API which I am providing. Such functionality is only difficult to maintain using Gson (own serializer, etc.)
And importing multiple JSON libraries is really not the idea of any project.

**Comment by aresot** [comment]
+1 I also need to use Jackson since we use getter/setter to de/serialize and Gson supports only property de/serialization. I tried following:
`
        for (SearchResult.Hit<String, Void> hit : topHits.getHits(String.class)) {
          Student obj = mapper.readValue(hit.source, Student.class);
          res.add(obj);
        }
` but JestResult.createSourceObject(...) will fill source only if object has fields and will return null. So there is no obvious way to extract source as String. Would be nice if getHits(String.class) or getHits(null) will fill source with json String.

**Comment by jmlw** [comment]
+1 for providing a configurable interface for deserialization. 

A real world use case for my company is adding results of a large computation producing a very large POJO with many nested objects, serialized with Jackson. When attempting to deserialize the json result of _source, enum deserialization fails because of different serialization / deserialization strategies. In my case, Gson failed silently with a duplicate key exception from deserializing an enum to null, and attempting to add these null keys to a map, returning a null value for my POJO. Being able to supply a Jackson object mapper rather than using the default Gson implementation will allow much greater flexibility.

For the time being, I've worked around this issue by extracting the JsonObject _source directly from the hits array with the json methods provided by Gson, then mapping each source to the POJO it represents.

**Comment by ssozonoff** [comment]
Looks like this has stalled for now. Seems GSON does not work well with Lombok either when deserialising something with an @Data annotation. Jackson just works but cant be plugged into Jest.
